command/agent/exec/exec_test.go

```diff
@@ -0,0 +1,288 @@
+			simulatedShutdownWaitDuration: 3 * time.Second,
+			expectedTestDuration:          15 * time.Second,
+		},
+
+		"test_app_ignores_stop_signal": {
+			envTemplates: []*ctconfig.TemplateConfig{{
+				Contents:                 pointerutil.StringPtr(`{{ with secret "kv/my-app/creds" }}{{ .Data.data.user }}{{ end }}`),
+				MapToEnvironmentVariable: pointerutil.StringPtr("MY_USER"),
+			}},
+			testAppArgs:                   []string{"--stop-after", "60s", "--sleep-after-stop-signal", "60s"},
+			testAppStopSignal:             syscall.SIGTERM,
+			testAppPort:                   34006,
+			simulatedShutdownWaitDuration: 32 * time.Second, // the test app should be stopped immediately after 30s
+			expectedTestDuration:          45 * time.Second,
+		},
+	}
+
+	goBinary, err := exec.LookPath("go")
+	if err != nil {
+		t.Fatalf("could not find go binary on path: %s", err)
+	}
+
+	// we must build a test binary since 'go run' does not propagate signals correctly
+	testAppBinary := filepath.Join(os.TempDir(), "test-app")
+
+	if err := exec.Command(goBinary, "build", "-o", testAppBinary, "./test-app").Run(); err != nil {
+		t.Fatalf("could not build the test application: %s", err)
+	}
+	defer func() {
+		if err := os.Remove(testAppBinary); err != nil {
+			t.Fatalf("could not remove %q test application: %s", testAppBinary, err)
+		}
+	}()
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			ctx, cancelContextFunc := context.WithTimeout(context.Background(), testCase.expectedTestDuration)
+			defer cancelContextFunc()
+
+			testAppCommand := []string{
+				testAppBinary,
+				"--port",
+				strconv.Itoa(testCase.testAppPort),
+			}
+
+			execServer := NewServer(&ServerConfig{
+				Logger: logging.NewVaultLogger(hclog.Trace),
+				AgentConfig: &config.Config{
+					Vault: &config.Vault{
+						Address: fakeVault.URL,
+						Retry: &config.Retry{
+							NumRetries: 3,
+						},
+					},
+					Exec: &config.ExecConfig{
+						RestartOnSecretChanges: "always",
+						Command:                append(testAppCommand, testCase.testAppArgs...),
+						RestartStopSignal:      testCase.testAppStopSignal,
+					},
+					EnvTemplates: testCase.envTemplates,
+				},
+				LogLevel:  hclog.Trace,
+				LogWriter: hclog.DefaultOutput,
+			})
+
+			// start the exec server
+			var (
+				execServerErrCh   = make(chan error)
+				execServerTokenCh = make(chan string, 1)
+			)
+			go func() {
+				execServerErrCh <- execServer.Run(ctx, execServerTokenCh)
+			}()
+
+			// send a dummy token to kick off the server
+			execServerTokenCh <- "my-token"
+
+			testAppHealthCheckCh := make(chan error)
+			testAppAddress := fmt.Sprintf("http://localhost:%d", testCase.testAppPort)
+
+			// check to make sure the test app is running
+			if testCase.expectedError == nil {
+				go func() {
+					time.Sleep(3 * time.Second)
+					_, err := http.Head(testAppAddress)
+					testAppHealthCheckCh <- err
+				}()
+			}
+
+			time.Sleep(5 * time.Second)
```

Comment:
I'm not sure if there's a way for us avoid this sleep as I'm not sure what it's for, but at the very least I think there should be a comment explaining why we have this sleep

Reply:
Agreed, I think this was a lazy way to have the internal goroutines start up, I'll see if it's still needed.

Reply:
I've removed this sleep, it seemed to be unnecessary.